### PR TITLE
Tarantool config storage

### LINKED ISSUE #3662: [Config] Document the config module API
**Product:** Tarantool
**Since:** 3.0
**Root document:** https://www.tarantool.io/en/doc/latest/reference/reference_lua/config/
**SME:** @ Totktonada

### LINKED ISSUE #3849: Create a diagram that reflects how a cluster's configuration is applied to instances
**Product:** Tarantool
**Since:** 3.0
**Root document:** https://docs.d.tarantool.io/en/doc/3.0/concepts/configuration/

# Details

For example, the most simple diagram might look as follows:

Local:
```
 _________________ _________________ _____
|                 |                 |     |
| Server 1        | Server   2      | ... |
|_________________ _________________ _____|
| router          | router          |     |
|_________________ _________________ _____|
| storage         | storag          |     |
|_________________ _________________ _____|
| storage         | storage         |     |
|_________________ _________________ _____|
| ...             | ...             |     |
|_________________|_________________|_____|
|     ____________|     ____________|     |
|    |cluster.yaml|    |cluster.yaml|     |
|____|____________|____|____________|_____|
```
Config storage:
```
 _________________ _________________ _____
|                 |                 |     |
| Server 1        | Server   2      | ... |
|_________________ _________________ _____|
| router          | router          |     |
|_________________ _________________ _____|
| storage         | storage         |     |
|_________________ _________________ _____|
| storage         | storage         |     |
|_________________ _________________ _____|
| ...             | ...             |     |
|_________________|_________________|_____|
         |            |          
         |____________|
                      | 
              ________|___________________
             | Centralized config      |  |
             |_________________________|  |
             |                            |        
             |   Configuration storage    |
             |____________________________|
```

We can also show additional entities (replica set, cluster) and demonstrate how specific configuration options apply to different instance types (router, storage).

### PR DESCRIPTION
**Pull request by andreyaksenov**
Updated topics:
- Updated intro: [Configuration](https://docs.d.tarantool.io/en/doc/tarantool-config-storage/concepts/configuration/)
- Updated the [Centralized configuration](https://docs.d.tarantool.io/en/doc/tarantool-config-storage/concepts/configuration/#centralized-configuration) section in `Configuration overview`.
- Reworked [Centralized configuration storages](https://docs.d.tarantool.io/en/doc/tarantool-config-storage/concepts/configuration/configuration_etcd/) to make it storage-agnostic. Added diagrams that demonstrate the difference between local and centralized configs.

New and updated samples [used](https://github.com/tarantool/doc/tree/latest/doc/code_snippets) in topics and reference pages:
1) Created a new [centralized_config](https://github.com/tarantool/doc/tree/tarantool-config-storage/doc/code_snippets/snippets/centralized_config/) folder for samples related to centralized configurations.
2) Moved the existing etcd sample (`etcd`) to this folder and renamed it to `config_etcd`.
3) Removed the `etcd_full` sample. The updated `config_etcd` sample now includes additional settings (auth and timeout).
4) Added the `etcd_config_storage.sh` script for quick setting up a protected etcd storage.
5) Created the `tarantool_config_storage` sample that shows how to set up a Tarantool-based storage.
6) Created the `config_storage` sample that shows how to connect to the Tarantool-based storage.
7) Added the `source.yaml` file to a tt environment directory. Can be used to test `tt cluster publish`.

image

What's new in 3.0:
- Updated `literalinclude` settings in a sample inside [Centralized configuration (EE)](https://docs.d.tarantool.io/en/doc/tarantool-config-storage/release/3.0.0/#centralized-configuration-ee).

Reference:
- Added a new [config.storage.*](https://docs.d.tarantool.io/en/doc/tarantool-config-storage/reference/configuration/configuration_reference/#config-storage) reference section.
- Added a new [roles](https://docs.d.tarantool.io/en/doc/tarantool-config-storage/reference/configuration/configuration_reference/#roles) section without descriptions. Will be documented in https://github.com/tarantool/doc/issues/3702.
- Documented the [config module API](https://docs.d.tarantool.io/en/doc/tarantool-config-storage/reference/reference_lua/config/) - it includes the `config.storage` API.
- Added sample to [conn:watch()](https://docs.d.tarantool.io/en/doc/tarantool-config-storage/reference/reference_lua/net_box/#conn-watch).

### CODE REVIEW

**Review by askalt — APPROVED** [review]
Thank you! lgtm

File: doc/concepts/configuration.rst [hunk 1]
```diff
@@ -327,25 +327,29 @@ Below are a few examples that show how to set environment variables of different
...
 
 ..  include:: /concepts/configuration/configuration_etcd.rst
-    :start-after: ee_note_etcd_start
-    :end-before: ee_note_etcd_end
+    :start-after: ee_note_centralized_config_start
+    :end-before: ee_note_centralized_config_end
+
 
-Tarantool enables you to store configuration data in one reliable place using `etcd <https://etcd.io/>`_.
+Tarantool enables you to store configuration data in one reliable place, for example, a Tarantool or etcd-based configuration storage.
```
Comment (p7nov) [inline doc/concepts/configuration.rst]: These are not `for example` - they are exact two supported ways of centralized config.  Let's explain this clear from the start.

File: doc/concepts/configuration.rst [hunk 2]
```diff
@@ -327,25 +327,29 @@ Below are a few examples that show how to set environment variables of different
...
 To achieve this, you need to:
 
-1.   Provide a local YAML configuration with an etcd endpoint address and key prefix in the ``config`` section:
+1.  Set up a centralized configuration storage.
+
+2.  Publish a cluster's configuration to the storage.
 
-    ..  literalinclude:: /code_snippets/snippets/config/instances.enabled/etcd/config.yaml
+3.  Configure a connection to the storage by providing a local YAML configuration with an endpoint address and key prefix in the ``config`` section:
```
Comment (p7nov) [inline doc/concepts/configuration.rst:345]: Nit: there can also be other connection parameters, maybe say it in a more general wording?

File: doc/concepts/configuration/configuration_etcd.rst [hunk 3]
```diff
@@ -1,94 +1,228 @@
 ..  _configuration_etcd:
```
Comment (p7nov) [inline doc/concepts/configuration/configuration_etcd.rst:1]: Should we change this anchor to reflect the new content?

File: doc/concepts/configuration/configuration_etcd.rst [hunk 4]
```diff
@@ -1,94 +1,228 @@
...
 
-.. ee_note_etcd_end
+.. ee_note_centralized_config_end
 
-Tarantool enables you to store configuration data in one place using `etcd <https://etcd.io/>`_.
-To achieve this, you need to define how to access etcd and publish a cluster's :ref:`YAML configuration <configuration_file>` to an etcd server.
+**Examples on GitHub**: `centralized_config <https://github.com/tarantool/doc/tree/latest/doc/code_snippets/snippets/centralized_config/>`_
 
+Tarantool enables you to store configuration data in one place using a Tarantool or etcd-based storage.
+With a :ref:`local YAML configuration <configuration_file>`, you need to make sure that all cluster instances use identical configuration files:
```
Comment (p7nov) [inline doc/concepts/configuration/configuration_etcd.rst:22]: I don't really like the explanation about the YAML file and the diagram with files on this page. Its a practical information for those who use local file config, while this page is not for them.
I think the proper place for the first diagram is the [Configuration in a file](https://docs.d.tarantool.io/en/doc/tarantool-config-storage/concepts/configuration/#configuration-in-a-file) section. 
The centralized diagram can be used both on the general Configuration ([Centralized section](https://docs.d.tarantool.io/en/doc/tarantool-config-storage/concepts/configuration/#centralized-configuration)) page and here.

File: doc/concepts/configuration/tarantool_config_centralized.png
Comment (p7nov) [inline doc/concepts/configuration/tarantool_config_centralized.png:1]: Why is the top block `Configuration servers` is left-aligned? I think the pic will look better with this block centered.

File: doc/concepts/configuration/tarantool_config_centralized_tcm.png
Comment (p7nov) [inline doc/concepts/configuration/tarantool_config_centralized_tcm.png:1]: It this diagram used somewhere? Didn't see it on pages.

File: doc/concepts/configuration/configuration_etcd.rst [hunk 5]
```diff
@@ -1,94 +1,228 @@
...
 
-.. _etcd_local_configuration:
+.. image:: tarantool_config_local.png
+    :align: left
+    :width: 500
+    :alt: Local configuration file
+
+Using a centralized configuration storage, all instances get the actual configuration from one place:
```
Comment (p7nov) [inline doc/concepts/configuration/configuration_etcd.rst:33]: Offsets look too small:
image

File: doc/concepts/configuration/configuration_etcd.rst [hunk 6]
```diff
@@ -1,94 +1,228 @@
...
+
+.. image:: tarantool_config_centralized.png
+    :align: left
+    :width: 500
+    :alt: Centralized configuration storage
+
+This topic describes how to set up a configuration storage, publish a cluster configuration to this storage, and use this configuration for all cluster instances.
```
Comment (p7nov) [inline doc/concepts/configuration/configuration_etcd.rst:44]: I feel the lack of conceptual explanation about storage types in the intro part. How do I choose before proceeding to practical steps?

What is Tarantool storage in regard to configurations? Does it include my cluster nodes? Are there specific requirements?
A short explanation about etcd? A link to its site/docs?

File: doc/concepts/configuration/configuration_etcd.rst [hunk 7]
```diff
@@ -1,94 +1,228 @@
...
+
+Setting up a configuration storage
+----------------------------------
+
+.. _centralized_configuration_storage_set_up_tarantool:
+
+Tarantool-based storage
+~~~~~~~~~~~~~~~~~~~~~~~
+
+A Tarantool-based storage is a replica set that stores configuration data in :ref:`synchronous <repl_sync>` spaces.
```
Comment (p7nov) [inline doc/concepts/configuration/configuration_etcd.rst]: > A Tarantool-based storage

I'm afraid that this wording it search-unfriendly, because `Tarantool storage` is a very common wording in docs but _outside_ the configuration context.
`Tarantool-based configuration storage` is indeed long name but I think we should use it at least in the beginning of each section so that readers know what they read about.

File: doc/concepts/configuration/configuration_etcd.rst [hunk 8]
```diff
@@ -1,94 +1,228 @@
...
+        :start-at: groups:
+        :end-at: 127.0.0.1:4403
+        :dedent:
+
+2.  :ref:`Create a user <configuration_credentials_managing_users_roles>` and grant them the following privileges:
+
+    *   The ``read`` and ``write`` permissions to the ``config_storage`` and ``config_storage_meta`` spaces used to store configuration data.
+    *   The ``execute`` permission to ``universe`` to allow interacting with the storage using the tt utility.
+
+    ..  literalinclude:: /code_snippets/snippets/centralized_config/instances.enabled/tarantool_config_storage/config.yaml
```
Comment (p7nov) [inline doc/concepts/configuration/configuration_etcd.rst:84]: Let's maybe use a self-explanatory username? `config_user` or something similar.

File: doc/concepts/configuration/configuration_etcd.rst [hunk 9]
```diff
@@ -1,94 +1,228 @@
...
+
+    $ tt start tarantool_config_storage
+
+Learn more from the :ref:`Starting and stopping instances <admin-start_stop_instance>` section.
+
+
+
+.. _centralized_configuration_storage_tarantool_interact:
+
+Interacting with the storage
```
Comment (p7nov) [inline doc/concepts/configuration/configuration_etcd.rst]: Not sure why this section is here. I expected the section about setting up etcd storage right here. 
This content seems relevant after at least one config is published to the storage.

File: doc/concepts/configuration/configuration_etcd.rst [hunk 10]
```diff
@@ -1,94 +1,228 @@
...
 
 
 
-.. _etcd_publishing_configuration:
+.. _centralized_configuration_storage_publish_config:
 
-Publishing a cluster's configuration to etcd
---------------------------------------------
+Publishing a cluster's configuration
+------------------------------------
```
Comment (p7nov) [inline doc/concepts/configuration/configuration_etcd.rst:155]: An intro sentence about ways to publish?

File: doc/concepts/configuration/configuration_etcd.rst [hunk 11]
```diff
@@ -1,94 +1,228 @@
...
+    The configured instances are used by tt when :ref:`starting a cluster <etcd_starting_instances>`.
+    ``tt cluster publish`` ignores this configuration file.
 
-To publish a cluster's configuration (``cluster.yaml``) to an etcd server, execute ``tt cluster publish`` as follows:
+To publish a cluster's configuration (``source.yaml``) to a centralized storage, execute ``tt cluster publish`` as follows:
 
 .. code-block:: console
 
-    $ tt cluster publish "http://localhost:2379/example" instances.enabled/app/cluster.yaml
+    $ tt cluster publish "http://localhost:2379/myapp" source.yaml
```
Comment (p7nov) [inline doc/concepts/configuration/configuration_etcd.rst]: I think that credentials are required in most cases, so let's maybe show them to save readers some time on reading the tt cluster reference?

File: doc/concepts/configuration/configuration_etcd.rst [hunk 12]
```diff
@@ -1,94 +1,228 @@
...
 
-    You can see a cluster's configuration using the :ref:`tt cluster show <tt-cluster>` command.
+    You can see a cluster's configuration using the ``tt cluster show`` command.
 
 
-.. _etcd_publishing_configuration_etcdctl:
+.. _centralized_configuration_storage_publish_config_etcdctl:
 
 Publishing configuration using etcdctl
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (p7nov) [inline doc/concepts/configuration/configuration_etcd.rst:221]: This maybe obvious but I'd start this section from an intro that it's only for etcd.

Comment (p7nov) [reply doc/concepts/configuration/configuration_etcd.rst:221]: Because in the previous section we said that there is no difference for tt between etcd and Tarantool storage :)
Somebody can be confused.

File: doc/concepts/configuration/configuration_etcd.rst [hunk 13]
```diff
@@ -106,21 +240,79 @@ To publish a cluster's configuration using the ``etcdctl`` utility, use the ``pu
 
 
 
+
+.. _centralized_configuration_storage_connect:
+
+Configuring connection to a storage
+-----------------------------------
+
+To use a centralized cluster's configuration, you need to provide connection settings in a local configuration file.
```
Comment (p7nov) [inline doc/concepts/configuration/configuration_etcd.rst]: This is where I would put a thought from your intro: despite we are talking about the centralized config, it's the reader's job to keep these files consistent and available on all cluster hosts.

**Review by p7nov — CHANGES_REQUESTED** [review]
Good job; some notes from my side regarding lack of explanations for some conceptual and specific things.

File: doc/concepts/configuration.rst (hunk 2 above)
Comment (andreyaksenov) [reply doc/concepts/configuration.rst:345]: This sample shows this minimum configuration and works more like a demo version of a centralized storage.

File: doc/concepts/configuration/tarantool_config_centralized.png
Comment (andreyaksenov) [reply doc/concepts/configuration/tarantool_config_centralized.png:1]: To keep the space for the potential diagram with TCM/tt/other tools (for example, `tarantool_config_centralized_tcm.png`, which is not used yet).

File: doc/concepts/configuration/tarantool_config_centralized_tcm.png
Comment (andreyaksenov) [reply doc/concepts/configuration/tarantool_config_centralized_tcm.png:1]: I've added it to be able to quickly insert it in docs if required. It has the same layout and width as other diagrams in this topic and might be useful.

File: doc/concepts/configuration/configuration_etcd.rst (hunk 9 above)
Comment (andreyaksenov) [reply doc/concepts/configuration/configuration_etcd.rst]: - dropped the `Interacting with the storage` and `Watching configuration updates` sections
- moved the net.box-related content to a reference page 
- created a new `Publishing configuration using the ‘config’ module` section that acts as an entry point for the `config.storage` API

File: doc/concepts/configuration/configuration_etcd.rst (hunk 10 above)
Comment (andreyaksenov) [reply doc/concepts/configuration/configuration_etcd.rst:155]: I hope, section names are self-explanatory and the intro won't add any additional info.

File: doc/concepts/configuration/configuration_etcd.rst (hunk 8 above)
Comment (andreyaksenov) [reply doc/concepts/configuration/configuration_etcd.rst:84]: Hmm.. not sure: `config_user` might be perceived as a built-in user. `sampleuser` doesn't add unnecessary meaning.

File: doc/concepts/configuration/configuration_etcd.rst [hunk 14]
```diff
@@ -1,103 +1,230 @@
...
+
+Setting up a configuration storage
+----------------------------------
+
+.. _centralized_configuration_storage_set_up_tarantool:
+
+Tarantool-based storage
+~~~~~~~~~~~~~~~~~~~~~~~
+
+To make a replica set act as a configuration storage, use the built-in ``config.storage`` :ref:`role <configuration_reference_roles_options>`.
```
Comment (p7nov) [inline doc/concepts/configuration/configuration_etcd.rst:58]: Can this replicaset be used for storing data at the same time? If not, this should be stated explicitly.

File: doc/concepts/configuration/configuration_etcd.rst (hunk 13 above)
Comment (p7nov) [reply doc/concepts/configuration/configuration_etcd.rst]: This reads as **centralized cluster's**. Maybe rephrase to
`To use a configuration from a centralized storage for your cluster...` ?

**Review by p7nov — APPROVED** [review]
Approved; just a couple more questions.

File: doc/concepts/configuration/configuration_etcd.rst (hunk 14 above)
Comment (andreyaksenov) [reply doc/concepts/configuration/configuration_etcd.rst:58]: I think, it can - it's just a regular replica set.
